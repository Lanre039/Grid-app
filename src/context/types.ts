export interface UserData {
  email: string;
  password: string;
  token: string;
}

export interface State {
  user: any;
}

export interface Action {
  type: string;
  payload?: any;
}
