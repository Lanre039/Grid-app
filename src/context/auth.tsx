import React, { createContext, useReducer } from "react";
import SecureLS from "secure-ls";
import { Action, State, UserData } from "./types";

const config = {
  encodingType: "des",
  isCompression: true,
  encryptionSecret: "my-secret-key",
};

// ENCRYPT THE DATA THEN SAVE TO LOCAL STORAGE
const ls = new SecureLS(config);
const localStore = {
  get: (item: string) => ls.get(item),
  set: (item: string, data = {}) => ls.set(item, data),
  remove: (item: string) => ls.remove(item),
  removeAll: () => ls.removeAll(),
};

const initialState: State = {
  user: null,
};

// CHECK IF USER HAS PREVIOUSLY LOGGED-IN
if (ls.get("userData")) {
  const userData = localStore.get("userData");
  initialState.user = userData;
}

const AuthContext = createContext({
  user: null,
  login: (user: UserData) => {},
  logout: () => {},
});

export function authReducer(state: State, action: Action) {
  switch (action.type) {
    case "LOGIN":
      return {
        ...state,
        user: action.payload,
      };
    case "LOGOUT":
      return {
        ...state,
        user: null,
      };
    default:
      return state;
  }
}

function AuthProvider(props: any) {
  const [state, dispatch] = useReducer(authReducer, initialState);

  function login(userData: UserData) {
    localStore.set("userData", userData);
    dispatch({
      type: "LOGIN",
      payload: userData,
    });
  }

  function logout() {
    localStore.removeAll();
    dispatch({
      type: "LOGOUT",
    });
  }

  return (
    <AuthContext.Provider
      value={{ user: state.user, login, logout }}
      {...props}
    ></AuthContext.Provider>
  );
}

export { AuthContext, AuthProvider, localStore };
