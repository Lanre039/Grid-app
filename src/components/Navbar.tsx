import { useContext } from "react";
import { AuthContext } from "../context/auth";
import { UserData } from "../context/types";
import userData from "../data/user.json";
import "./navbar.scss";

function Navbar() {
  const { user, login, logout } = useContext(AuthContext);

  const loginUser = () => login(userData as UserData);
  const logOutUser = () => logout();

  const handleClick = () => {
    if (user) {
      logOutUser();
    } else {
      loginUser();
    }
  };
  return (
    <nav className="nav">
      <div className="nav_content container">
        <h1>SHAPES</h1>
        <h1 className="nav_text-right" onClick={handleClick}>
          {user ? "Logout" : "Login"}
        </h1>
      </div>
    </nav>
  );
}

export default Navbar;
