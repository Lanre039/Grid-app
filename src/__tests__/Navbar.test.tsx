import { cleanup } from "@testing-library/react";
import { mount } from "enzyme";
import { shallow } from "enzyme";
import Navbar from "../components/Navbar";
import { AuthContext } from "../context/auth";

afterEach(cleanup);

const auth = {
  user: null,
  login: jest.fn(),
  logout: jest.fn(),
};

describe("Navbar", () => {
  it("render correctly", () => {
    const wrapper = shallow(<Navbar />);
    expect(wrapper).toMatchSnapshot();
  });

  it("login user", () => {
    const screen = mount(
      <AuthContext.Provider value={auth}>
        <Navbar />
      </AuthContext.Provider>
    );

    const loginBtn = screen.find(".nav_text-right");
    loginBtn.simulate("click");
    expect(auth.login).toHaveBeenCalledTimes(1);
  });
});
