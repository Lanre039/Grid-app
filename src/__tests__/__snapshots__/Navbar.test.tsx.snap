// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar render correctly 1`] = `
<nav
  className="nav"
>
  <div
    className="nav_content container"
  >
    <h1>
      SHAPES
    </h1>
    <h1
      className="nav_text-right"
      onClick={[Function]}
    >
      Login
    </h1>
  </div>
</nav>
`;
